gtk4/src/subclass/widget.rs

```diff
@@ -123,3 +126,127 @@ unsafe extern "C" fn widget_direction_changed<T: ObjectSubclass>(
+        klass: &mut subclass::simple::ClassStruct<Self>,
+        template: &glib::Bytes,
+    ) {
+        let type_class = klass as *mut _ as *mut gobject_sys::GTypeClass;
+        let widget_class =
+            gobject_sys::g_type_check_class_cast(type_class, gtk_sys::gtk_widget_get_type())
+                as *mut gtk_sys::GtkWidgetClass;
+        gtk_sys::gtk_widget_class_set_template(widget_class, template.to_glib_none().0);
+    }
+
+    unsafe fn set_template(klass: &mut subclass::simple::ClassStruct<Self>, template: &[u8]) {
+        let template_bytes = glib::Bytes::from(template);
+        Self::set_template_bytes(klass, &template_bytes);
+    }
+
+    unsafe fn set_template_static(
+        klass: &mut subclass::simple::ClassStruct<Self>,
+        template: &'static [u8],
+    ) {
+        let template_bytes = glib::Bytes::from_static(template);
+        Self::set_template_bytes(klass, &template_bytes);
+    }
+
+    unsafe fn set_template_from_resource(
+        klass: &mut subclass::simple::ClassStruct<Self>,
+        resource_name: &str,
+    ) {
+        let type_class = klass as *mut _ as *mut gobject_sys::GTypeClass;
+        let widget_class =
+            gobject_sys::g_type_check_class_cast(type_class, gtk_sys::gtk_widget_get_type())
+                as *mut gtk_sys::GtkWidgetClass;
+        gtk_sys::gtk_widget_class_set_template_from_resource(
+            widget_class,
+            resource_name.to_glib_none().0,
+        );
+    }
+
+    unsafe fn bind_template_child(klass: &mut subclass::simple::ClassStruct<Self>, name: &str) {
+        let type_class = klass as *mut _ as *mut gobject_sys::GTypeClass;
+        let widget_class =
+            gobject_sys::g_type_check_class_cast(type_class, gtk_sys::gtk_widget_get_type())
+                as *mut gtk_sys::GtkWidgetClass;
+        gtk_sys::gtk_widget_class_bind_template_child_full(
+            widget_class,
+            name.to_glib_none().0,
+            false as glib_sys::gboolean,
+            0,
+        );
+    }
+
+    unsafe fn bind_template_child_with_offset<T>(
+        klass: &mut subclass::simple::ClassStruct<Self>,
+        name: &str,
+        offset: field_offset::FieldOffset<Self, TemplateWidget<T>>,
+    ) where
+        T: ObjectType + FromGlibPtrNone<*mut <T as ObjectType>::GlibType>,
+    {
+        let type_class = klass as *mut _ as *mut gobject_sys::GTypeClass;
+        let widget_class =
+            gobject_sys::g_type_check_class_cast(type_class, gtk_sys::gtk_widget_get_type())
+                as *mut gtk_sys::GtkWidgetClass;
+        let private_offset = Self::type_data().as_ref().private_offset;
+        gtk_sys::gtk_widget_class_bind_template_child_full(
+            widget_class,
+            name.to_glib_none().0,
+            false as glib_sys::gboolean,
+            private_offset + (offset.get_byte_offset() as isize),
+        )
+    }
+}
+
+#[derive(Debug)]
+#[repr(transparent)]
+pub struct TemplateWidget<T>
```

Comment:
This may work for more than widgets - perhaps `TemplateChild` would be more appropriate.

Reply:
Can `gtk::Builder` be used for anything but widgets?

Reply:
Vala uses `[GtkChild]` annotation while Python uses `Gtk.Template.Child()`